do not allow using crr location as a locationConstraint

welansari opened the pull request:
- Support the 'CRR' type for locations
- Do not allow setting the CRR location as a locationConstraint

Issue: CLDSRV-653

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-653 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.12`

* `9.1.0`


Please check the `Fix Version/s` of CLDSRV-653, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-653 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.0`


Please check the `Fix Version/s` of CLDSRV-653, or the target
branch of this pull request.

francoisferrand commented on lib/api/bucketPut.js:
```diff
@@ -59,7 +59,8 @@ function checkLocationConstraint(request, locationConstraint, log) {
     }
-    if (locationConstraints[locationConstraintChecked].isCold) {
+    if (locationConstraints[locationConstraintChecked].isCold ||
+        locationConstraints[locationConstraintChecked].type === 'crr') {
```
instead of doing the check based on the `type`, maybe we could to introduce a `feature` parameter, like isCold → `isCrr` ?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/api/bucketPut.js:
```diff
@@ -59,7 +59,8 @@ function checkLocationConstraint(request, locationConstraint, log) {
             customizeDescription(errMsg) };
     }
-    if (locationConstraints[locationConstraintChecked].isCold) {
+    if (locationConstraints[locationConstraintChecked].isCold ||
```
location constraint can be set on bucket (and should thus be rejected), but can also be specified directly in putObject / copyObject / putMpu : this needs to be handled as well

francoisferrand replied:
it seems weird that we don't have a similar check on `isCrr` in the other route (objectPut, `initiateMultiPartIUpload`, `copyObject` - or a common function called by these), to explicitly detect (and reject) the write to a crr location.
how come the test succeed? Is there some filtering somewhere?

welansari replied:
We don't do any filtering in those, we directly use the locationConstraint check function defined in Arsenal. We call
https://github.com/scality/cloudserver/blob/development/9.0/lib/api/apiUtils/object/locationConstraintCheck.js at some point in the code to get the controlling locationConstraint or fail when it's CRR. PutObject runs the check quite late (we do it in `createAndStoreObject`)

Commit:
support crr location in the locationConstaint config list

Issue: CLDSRV-653

Commit:
do not allow setting crr location as a bucket location constraint

Issue: CLDSRV-653

Commit:
do not allow crr location as an object locationConstraint

Issue: CLDSRV-653

williamlardier commented on tests/multipleBackend/backendHealthcheckResponse.js:
```diff
@@ -22,8 +22,8 @@ describe('Healthcheck response', () => {
-                if (constraint === 'location-dmf-v1') {
-                    // FIXME: location-dmf-v1 is not in results, see CLDSRV-440
+                if (constraint === 'location-dmf-v1' || constraint === 'location-crr-v1') {
+                    // FIXME: location-dmf-v1 and location-crr-v1 are not in results, see CLDSRV-440
```
Maybe we shoudl update the ticket to include CRR now?

williamlardier commented on lib/Config.js:
```diff
@@ -385,6 +385,10 @@ function locationConstraintAssert(locationConstraints) {
             assert(locationConstraints[l].isCold === true);
         }
 
+        if (locationConstraints[l].type === 'crr') {
+            assert(locationConstraints[l].isCrr === true);
+        }
```
If we always have the "type" crr and expect isCRR to be true, why not just relying on the `type` to detect if this is a CRR location instead?

francoisferrand replied:
the point is actually the opposite: there may be multiple "types" of locations which support CRR (i.e. must be set in backbeat replication endpoint) but cannot be used for writing directly. So the flag `isCrr` was introduced to give this flexibility; and type may be used if we have different kinds of crr locations.

francoisferrand commented on lib/Config.js:
```diff
@@ -385,6 +385,10 @@ function locationConstraintAssert(locationConstraints) {
             assert(locationConstraints[l].isCold === true);
         }
 
+        if (locationConstraints[l].type === 'crr') {
```
why do we want to check this?
The code should use `isCrr` to check if this is a car-only location, so we should not check for the type? (type could actually be scality, in that case)

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/CLDSRV-653` and the
destination branch `development/9.1`.

Please resolve the conflict on **the feature branch** (`improvement/CLDSRV-653`).

```sh
git fetch && \
git checkout origin/improvement/CLDSRV-653 && \
git merge origin/development/9.1
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/CLDSRV-653
```

francoisferrand commented on tests/unit/api/objectCopy.js:
```diff
@@ -338,6 +339,22 @@ describe('non-versioned objectCopy', () => {
+            next => objectCopy(authInfo, testObjectCopyRequest, sourceBucketName, objectKey,
+                undefined, log, next),
+        ], err => {
+            assert(err.is.InvalidArgument);
```
is this the right error?
should we not sent `InvalidLocationConstraint` in that case (i.e. trying to write on a valid location, but which does not support writing) instead of `InvalidArgument` (simply because we have no backend class for the type `crr` I guess?)

welansari replied:
Looking at this test, it seems like there was a reason behind returning this error, not sure why however.
https://github.com/scality/cloudserver/blob/96a0315a8b7d436e4945630b4c0821bf5e7480aa/tests/unit/multipleBackend/locationConstraintCheck.js#L37-L46

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

williamlardier commented on tests/unit/api/multipartUpload.js:
```diff
@@ -2174,6 +2174,32 @@ describe('Multipart Upload API', () => {
+            objectKey,
+            headers: {
+                host: `${bucketName}.s3.amazonaws.com`,
+                [constants.objectLocationConstraintHeader]: 'location-crr-v1',
```
This is used multiple times, maybe better to add a global constant instead? (same for DMF?)

Commit:
replace hardcoded location names with global test variables

Replaced all instances of 'location-dmf-v1' and 'location-crr-v1'
in test files with references to global constants

Issue: CLDSRV-653

welansari commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`
* `development/9.0`



Please check the status of the associated issue CLDSRV-653.

Goodbye kerkesni.




*The following options are set:* **approve**
